File: Day 1/day-1.py
import requests
from day_1_puzzel_input import *


def errorHandler(list: list, index: int) -> bool:
    errorPrefix = f"The list:{list} at index: {index}"
    if len(list) == 0:
        raise ValueError(f"{errorPrefix} is empty")
    if not all(i.isnumeric() for i in list):
        raise ValueError(f"{errorPrefix} contains non-numeric values")
    if not all(int(i) < 10**6 for i in list):
        raise ValueError(f"{errorPrefix} contains values greater than 10^6")
    if len(list) > 2:
        raise ValueError(f"{errorPrefix} contains more than 2 items")
    return True


def parseInputStringToIntList(longString: str) -> dict[str, list[int]]:
    numsList: list[str] = longString.split("\n")
    list1: list[int] = []
    list2: list[int] = []
    for index, value in enumerate(numsList):
        if value == "":
            continue
        numTuple = value.split()
        errorHandler(numTuple, index)
        list1.append(int(numTuple[0]))
        list2.append(int(numTuple[1]))
    list1.sort()
    list2.sort()
    return {"listOne": list1, "listTwo": list2}


def getListDifference(listTuple: dict[str, list[int]]) -> int:
    list1, list2 = listTuple.values()
    difference = 0
    if len(list1) != len(list2):
        raise ValueError("The two lists are not of the same length")
    for i in range(len(list1)):
        difference += abs(list1[i] - list2[i])
    print("The difference between the two lists is: ", difference)
    return difference


getListDifference(parseInputStringToIntList(longString))


def createDictWithDuplicateCount(listTwo: list[int]) -> dict[int, int]:
    countedDuplicates = {}
    for i in listTwo:
        if i in countedDuplicates:
            countedDuplicates[i] += 1
        else:
            countedDuplicates[i] = 1
    return countedDuplicates


def getSimilarityScore(listOne: list[int], countedDuplicates: dict[int, int]) -> int:
    score = 0
    for i in listOne:
        if i in countedDuplicates:
            score += i * countedDuplicates[i]
    print("The similarity score is: ", score)
    return score


def getSimilarityScoreFromInputString(longString: str) -> int:
    parsedInput = parseInputStringToIntList(longString)
    listOne, listTwo = parsedInput.values()
    countedDuplicates = createDictWithDuplicateCount(listTwo)
    return getSimilarityScore(listOne, countedDuplicates)


getSimilarityScoreFromInputString(longString)

File: Day 1/day_1.js
const fs = require("fs");

function getInput() {
  const inputList = fs
    .readFileSync(
      "C:/Users/RickLyman/Desktop/Coding Challenges/2024-CodingAdvent Challenges/Day 1/day_1_puzzel_input.txt",
      "utf-8"
    )
    .trim()
    .split("\r\n")
    .map((line) => line.split(/\s+/).map(Number));
  const list1 = inputList.map((pair) => pair[0]).sort((a, b) => a - b);
  const list2 = inputList.map((pair) => pair[1]).sort((a, b) => a - b);
  return [list1, list2];
}

function getDifferences(list1, list2) {
  if (list1.length !== list2.length)
    throw new Error("Lists are not the same length");
  return list1.reduce((total, num, i) => total + Math.abs(num - list2[i]), 0);
}

function getSimilarityScore(list1, list2) {
  const duplicateCounts = list2.reduce((acc, num) => {
    acc[num] = (acc[num] || 0) + 1;
    return acc;
  }, {});

  return list1.reduce(
    (score, num) => score + (duplicateCounts[num] || 0) * num,
    0
  );
}
const [list1, list2] = getInput();
console.log(getDifferences(list1, list2));
console.log(getSimilarityScore(list1, list2));
